Added tests for mkdir_syscall

namanlalitnyu opened the pull request:
## Description

Fixes # (issue)
The following changes include the tests for the "mkdir" file system call under RustPosix.
The tests were added to cover all the possible scenarios that might happen when calling the file system_call `mkdir_syscall`. 

### Type of change
- [ ]  This change just contains the tests for an existing file system call.

## How Has This Been Tested?
Inorder to run the tests, we need to run `cargo test --lib` command inside the `safeposix-rust` directory.

All the tests are present under this directory: `lind_project/src/safeposix-rust/src/tests/fs_tests.rs`

- Test A - `ut_lind_fs_mkdir_empty_directory()`
- Test B - `ut_lind_fs_mkdir_nonexisting_directory()`
- Test C - `ut_lind_fs_mkdir_existing_directory()`
- Test D - `ut_lind_fs_mkdir_invalid_modebits()`
- Test E - `ut_lind_fs_mkdir_success()`
- Test F - `ut_lind_fs_mkdir_using_symlink()`

## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

namanlalitnyu committed:
Added tests for mkdir_syscall

Yaxuan-w reviewed (approved):
Overall logic makes sense to me, lgtm so approved

JustinCappos reviewed (commented):
Looks great!  Can you add a test for doing mkdir through a symlink?

yashaswi2000 commented:
Great starter! i think we can bundle the comments update for the syscall in this PR itself.

namanlalitnyu committed:
Added symlink test for mkdir_syscall

namanlalitnyu commented:
> Looks great! Can you add a test for doing mkdir through a symlink?

Added the test for the same.

JustinCappos commented on src/tests/fs_tests.rs:
```diff
@@ -1237,6 +1245,81 @@ pub mod fs_tests {
+        lindrustinit(0);
+        let cage = interface::cagetable_getref(1);
+
+        // Create a directory which will be referred to as originaldir 
+        let fd = cage.open_syscall("/originaldir", O_CREAT | O_EXCL | O_WRONLY, S_IRWXA);
+        assert_eq!(cage.write_syscall(fd, str2cbuf("hi"), 2), 2);
```
This creates a file, not a directory, right?

namanlalitnyu replied:
> This creates a file, not a directory, right?
Sorry, my bad...I just rechecked the other references for this function call and it creates a file, not a directory. Updated the changes.

namanlalitnyu committed:
Updated symlink test for mkdir_syscall

namanlalitnyu committed:
Updated fs_mkdir call with comments and updated test

namanlalitnyu commented on src/safeposix/syscalls/fs_calls.rs:
```diff
@@ -221,32 +223,37 @@ impl Cage {
                     parentdir.linkcount += 1;
+                    parentdir.ctime = time;
+                    parentdir.mtime = time;
+                    // Here, update the ctime and mtime for the parent directory as well
```
Man pages for this call contain the update to the timestamps as well - https://docs.oracle.com/cd/E86824_01/html/E54765/mkdir-2.html

namanlalitnyu commented on src/tests/fs_tests.rs:
```diff
@@ -1237,6 +1245,98 @@ pub mod fs_tests {
+        assert_eq!(cage.mkdir_syscall("/parentdir/dir", S_IRWXA), 0);
+        
+        // Get the stat data for the child directory and check for inode link count to be 3 initially
+        let mut statdata2 = StatData::default();
```
Updated this test to verify the "link count" when a new directory is created.

JustinCappos commented on src/safeposix/syscalls/fs_calls.rs:
```diff
@@ -187,6 +187,8 @@ impl Cage {
             return syscall_error(Errno::ENOENT, "mkdir", "given path was null");
         }
+
+        // Get the absolute path
```
Why do we do this?

namanlalitnyu replied:
This is done to get the absolute path from the root directory of Lind so that we can trace the entire path and check for errors.

JustinCappos replied:
okay, please explain why we need to do this to check for errors in the code comment. 

I can tell you're getting the absolute path from the code, but don't know why you are getting it until you tell me...

namanlalitnyu replied:
Got it, will update the comments.

namanlalitnyu replied:
Updated the comments.

JustinCappos commented on src/safeposix/syscalls/fs_calls.rs:
```diff
@@ -221,32 +223,37 @@ impl Cage {
                 }
+                // Insert the newly formed inode to the inodetable
                 metadata.inodetable.insert(newinodenum, newinode);
+                // Update the logs with metadata
```
Why?  What is being said here?

namanlalitnyu replied:
I think we can remove this comment as the line below it is self-explanatory.

JustinCappos reviewed (commented):
Looks pretty good.  I like the tests particularly.  A few comments could be improved, but it mostly looks ready.

yashaswi2000 commented:
A brief intro and info about the arguments, Flags being used, purpose and return types above the function signature will make the documentation whole. you can refer to the man pages to add this, potentially also include any limitations and unsupported features we have today as a Note. 

other than that, everything looks good!

namanlalitnyu committed:
Added information as comments

namanlalitnyu commented:
> A brief intro and info about the arguments, Flags being used, purpose and return types above the function signature will make the documentation whole. you can refer to the man pages to add this, potentially also include any limitations and unsupported features we have today as a Note.
> 
> other than that, everything looks good!


Done, added relevant information before the system call and added more context inside the function call for better understanding. Couldn't find any limitations / unsupported features for now, but will update it once I find in future.

yashaswi2000 reviewed (approved):
LGTM!
